test/integration/proxy_test.exs

```diff
@@ -124,4 +124,18 @@ defmodule Supavisor.Integration.ProxyTest do
     assert %P.Result{num_rows: 0} =
              P.query!(origin, "select * from public.test where details = 'test_delete'", [])
   end
+
+  test "too many clients in session mode" do
+    db_conf = Application.get_env(:supavisor, Repo)
+
+    url =
+      "postgresql://session.proxy_tenant:#{db_conf[:password]}@#{db_conf[:hostname]}:#{Application.get_env(:supavisor, :proxy_port)}/postgres"
+
+    spawn(fn -> System.cmd("psql", [url], stderr_to_stdout: true) end)
+
+    :timer.sleep(500)
```

Comment:
Right, we need to wait to make sure the client connection is up. Another way of solving this would be to allow hooking into the proxy lifecycle. For example, with Postgrex (thanks to the underlying DBConnection) we can do this:

```elixir
pid = self()

{:ok, _} =
  Postgrex.start_link(..., after_connect: fn _conn ->
    send(pid, :connected)
  end)

assert_receive :connected, 500
```

which is really convenient in tests. Perhaps it's something worth considering.

Although in this particular case it could be a bit difficult to hook into given it seems we're not explicitly starting proxy and instead we rely on the one started in Supavisor's application supervision tree and so the only way to configure it would be using application env which then would affect other tests. So yeah unfortunately it's not trivial but perhaps worth pursuing after all in other PRs.

Reply:
Thanks, that is very interesting. I will try it in a separate commit 👍